Drop caps

Stwerp opened the pull request:
In the books, chapters tend to start with a drop capital in a semi-blackletter font. I created a command to start a new paragraph (and updated the example.tex). It uses fonts that come with TeXLive distribution and does not require xelatex or fontspec. (Command naming comes from IEEEParStart that is used for IEEE publications). A new requirements is the lettrine package (and font packages).

Usage:
    \dndParStart[<optional: extra lines>]{<first character>}{<rest of line>} ... rest of paragraph
The first mandatory argument is the letter to be drop-capital'ed. The 2nd argument is the first line which is set in small caps. It takes trial and error to determine where to end this argument (lettrine package will not play nicely with \FirstLine)
Example Usage:
    \dndParStart{L}{eave all your typesetting woes} behind you as you begin your journey.
Appearance:
![image](https://user-images.githubusercontent.com/1114823/54043542-dc66bb80-419a-11e9-90af-ec12211de567.png)

\dndParStart[1]{Y}{our party begins their journey} in typesetting.
Note the extra line is needed for the y descender.
![image](https://user-images.githubusercontent.com/1114823/54053895-b5b67e00-41b6-11e9-9398-71c91184b8f5.png)

There is also a new package option *altdropfont* that will set the letter in a more gothic font. For some letters, this font will appear closer to the source material. (Particularly, the “T” as shown below).
![image](https://user-images.githubusercontent.com/1114823/54053904-bb13c880-41b6-11e9-99cf-2a2c7292bf58.png)

BrianCriswell commented:
This is a great addition! There will need to be some adjustments made to have it fit a little better in the package, but I will walk you through it. :)

1. The continuous integration is failing because it does not know to include the `lettrine` package. Add "lettrine" to packages.txt in alphabetical order.
1. Add a short note regarding the new dropcap function to the unreleased section of `changelog.md`.
1. Adjust the small caps line so that it is vertically aligned with the lines underneath.

The rest of the items can be handled in code review.

Stwerp commented:
I was working on it, adding package: "ArtNouvc" and "lettrine"

This will align the text correctly (even left justification)
    \newcommand{\dndParStart}[3][0]{\lettrine[lines=4,depth=#1,findent=0.5em,nindent=0em]{\dndDropFont #2}{}\textsc{#3}}

This will use the correct package instead of the \input format
    \RequirePackage{ArtNouvc}
    \newcommand*\dndDropFont{\ArtNouvcfamily}

I can take a look at this more this weekend, and work on the naming. Thanks for the help! (this is my first pull request, and am not famaliar with the testing system!)

BrianCriswell commented:
> I was working on it, adding package: "ArtNouvc" and "lettrine"
> 
> This will align the text correctly (even left justification)
> \newcommand{\dndParStart}[3][0]{\lettrine[lines=4,depth=#1,findent=0.5em,nindent=0em]{\dndDropFont #2}{}\textsc{#3}}
> 
> This will use the correct package instead of the \input format
> \RequirePackage{ArtNouvc}
> \newcommand*\dndDropFont{\ArtNouvcfamily}
> 
> I can take a look at this more this weekend, and work on the naming. Thanks for the help! (this is my first pull request, and am not famaliar with the testing system!)

No worries. Sometimes we get PRs where the initial post is it with no further development.

Be careful with using `\RequirePackage` with fonts. Depending on the font style it belongs to, it will replace the family (serif, sans serif, etc.) with the new font. You most likely can use the font directly.

```latex
\newcommand*\dnd@DropCapFont{\fontfamily{ArtNouvc}\selectfont}
```

I look forward to how this develops. :)

Stwerp commented on dnd.sty:
```diff
@@ -27,6 +27,7 @@
 \RequirePackage{xcolor}
 \RequirePackage{xkeyval}
 \RequirePackage{xparse}
+\RequirePackage{lettrine}
```
I updated package order in the file

Stwerp commented on dnd.sty:
```diff
@@ -42,6 +43,10 @@
 % Options
 %
 
+% Allow selecting alternate drop-cap font
```
Agree. Removed the option, and instead am using the lettrine packages fontHooks. Users can issue a `\usepackage{tgothic}` and `\renewcommand{\LettrineFontHook}{\tgothfamily}` in their document preamble to alter the font.

Stwerp commented on dnd.sty:
```diff
@@ -153,6 +158,21 @@
+\newcommand\dndDropFont{\tgothfamily}
+}{% use default
+\input ArtNouvc.fd
+\newcommand*\dndDropFont{\usefont{U}{ArtNouvc}{xl}{n}}
```
I created a new dndfonts.sty file for this. It seems simple for a single `newcommand` issue, but if further font refinements happen, it would be the place for it.

BrianCriswell replied:
There already is a dndfonts.sty. Most of the sub-modules are in the `lib` folder.

Stwerp replied:
Oof. Ok, using the right file now. Would it also make more sense to define the `DnDDropCapLine` command in `lib/dndfonts.sty` as well?

Stwerp commented on dnd.sty:
```diff
@@ -153,6 +158,21 @@
+%    the linebreak. Lettrine package will not play nicely with \FirstLine
+%    The initial argument is optional and is used for letters with lower descender (default is 0)
+%    For example a capital Y needs an extra line
+\newcommand{\dndParStart}[3][0]{\lettrine[lines=4,depth=#1,findent=0.2em]{\dndDropFont #2}{}\textsc{#3}}
```
I updated to the new format, and named it `dndDropCapLine`. Other commands in the file have a lowercase `dnd` prefix.

BrianCriswell replied:
LaTeX document-level commands (especially in the xparse / LaTeX3 realm) are moving toward upper-camel case (Pascal case) with the first letter capitalized.

Stwerp commented:
My apologies for these commits! If you'd like to close this PR, I can rebase the commits from me fighting the CI tools, and re-submit for a code review. I didn't realize I could run circleCI myself nor that my pushes to this branch were being spammed around to you and everyone watching the repo.

Changes from `3420d9c` are:
+ Added line in `CHANGELOG.MD` to reflect new command
+ In `dnd.sty`: moved `RequirePackage{lettrine}` to alphabetical location
+ In `dnd.sty`: Drop Cap font is selected using the `LettrineFontHook` command provided by `lettrine.sty`
+ in `dnd.sty`: Created `dndDropCapLine` using `NewDocumentCommand` format
- in `dnd.cls`: Removed package option for 'altdropfont'
- in `dnd.cls`: removed package option 'altdropfont'
+ Added `dndfonts.sty` which defines the font `dnd@DropCapFont`

commits after that are configuring packages.txt to select the appropriate packages from tlmgr, and identifying the hanging newline character.

BrianCriswell commented:
> My apologies for these commits! If you'd like to close this PR, I can rebase the commits from me fighting the CI tools, and re-submit for a code review. I didn't realize I could run circleCI myself nor that my pushes to this branch were being spammed around to you and everyone watching the repo.
> 
> Changes from `3420d9c` are:
> 
> * Added line in `CHANGELOG.MD` to reflect new command
> * In `dnd.sty`: moved `RequirePackage{lettrine}` to alphabetical location
> * In `dnd.sty`: Drop Cap font is selected using the `LettrineFontHook` command provided by `lettrine.sty`
> * in `dnd.sty`: Created `dndDropCapLine` using `NewDocumentCommand` format
> 
> * in `dnd.cls`: Removed package option for 'altdropfont'
> * in `dnd.cls`: removed package option 'altdropfont'
> 
> * Added `dndfonts.sty` which defines the font `dnd@DropCapFont`
> 
> commits after that are configuring packages.txt to select the appropriate packages from tlmgr, and identifying the hanging newline character.

No worries, anyone can mute the notifications on a PR if they want to do so. The main thing is that it passes the test and follows our conventions.

BrianCriswell commented on lib/dndfonts.sty:
```diff
@@ -20,4 +20,4 @@
 
-\newcommand\dnd@DropCapFont{\ArtNouvcfamily}
\ No newline at end of file
+\newcommand\dnd@DropCapFont{\ArtNouvcfamily}
```
The check that the linting had failed on was referring to there being a blank line at the end of the file rather than inserting a `\n`. This is automatically supplied by .editorconfig-compatible editors on file save. Which editor are you using?

Stwerp replied:
I use emacs. It looks like emacs was eating my last line (I went to last line, killed any remaining text, and emacs would simply eat the entire line). I got an editorconfig package, and it looks like it is taking care of the whitespace issues. (Testing on two computers: linux and mac)

Stwerp commented:
Ok, I think this is all together now. Lint is happy and the CI toolchain is happy. Let me know what you think.

BrianCriswell commented:
I have not forgotten about this. Just a bit busy.

BrianCriswell commented on dnd.sty:
```diff
@@ -153,6 +153,15 @@
   \dndline%
 }
 
+\renewcommand{\LettrineFontHook}{\dnd@DropCapFont}
```
This is primarily font related and should probably be in `lib/dndfonts.sty`.

BrianCriswell commented on dnd.sty:
```diff
@@ -153,6 +153,15 @@
+%    The initial argument is optional and is used for letters with lower descender (default is 0)
+%    For example a capital Y needs an extra line
+\NewDocumentCommand{\DndDropCapLine}{ O{0} m m }%
+{\lettrine[lines=4,depth=#1,findent=0.5em,nindent=0em]{#2}{}\textsc{#3}}
```
`\lettrine` already puts the third argument in small caps, so we can remove the `\textsc`. Also, with a small change, we can expose all the power of `\lettrine` while setting the defaults we want. This is especially useful for finicky fonts like ArtNouveauC. `\wordsep` may look more natural for the gap on the right of the drop cap.

```
\NewDocumentCommand{\DndDropCapLine}{ O{} m m }%
{\lettrine[lines=4,depth=0,findent=\wordsep,nindent=0em,#1]{#2}{#3}}
```

BrianCriswell commented on lib/dndfonts.sty:
```diff
@@ -18,3 +19,5 @@
 \newcommand{\dnd@FooterFont}{\normalfont\scshape}
 \newcommand{\dnd@PageNumberFont}{\normalfont}
+
+\newcommand\dnd@DropCapFont{\ArtNouvcfamily}
```
ArtNouveauc is a nice font, but it is problematic as the default because of all the parts of different letters that extend to the right and downwards. I would propose using a different font, one that has the letters fit more within the box. http://www.tug.dk/FontCatalogue/otherfonts.html

RoyalInitialen is a possiblity, although the styling may not work for some, and most of the fonts are problematic in that regard: there is not really one that works for everybody's preference.

```
\input RoyalIn.fd
\newcommand*\RoyalInfamily{\usefont{U}{RoyalIn}{xl}{n}}
\newcommand\dnd@DropCapFont{\RoyalInfamily}
\renewcommand{\LettrineFontHook}{\dnd@DropCapFont}
```

BrianCriswell commented:
I made a few changes:
- Switched to Royal Initialen which does not have tails sticking out and should work with every letter with no modification
- Changed box titles and other uses of opensans to kp-fonts as opensans dropped support for small caps
- Exposed all of lettrine's functionality through our wrapper function.

I will leave this up for a few days in case anyone wants to review it.
